File: src/portfolio_opt/rebalance.py
from __future__ import annotations

from math import floor

from .config import OptimizationConfig
from .types import (
    AccountSnapshot,
    OrderPlan,
    Position,
    TrailingStopPlan,
    TrailingStopPlanResult,
    UnprotectedTrailingStopQty,
)


def current_weights(
    symbols: list[str],
    account: AccountSnapshot,
    positions: list[Position],
) -> dict[str, float]:
    by_symbol = {position.symbol: position for position in positions}
    if account.equity <= 0:
        raise ValueError("Account equity must be positive.")
    # Any symbol not currently held is treated as a zero-weight position so the
    # optimizer and rebalance layer can work off the same ordered universe.
    return {
        symbol: by_symbol.get(
            symbol, Position(symbol=symbol, qty=0.0, market_value=0.0)
        ).market_value
        / account.equity
        for symbol in symbols
    }


def build_order_plan(
    symbols: list[str],
    target_weights: list[float],
    account: AccountSnapshot,
    positions: list[Position],
    latest_prices: dict[str, float],
    config: OptimizationConfig,
    open_orders: list[dict] | None = None,
) -> list[OrderPlan]:
    weights_now = current_weights(symbols, account, positions)

    # Adjust current weights for any pending orders so we don't
    # double-submit or submit conflicting trades.
    if open_orders:
        for order in open_orders:
            symbol = order.get("symbol")
            if symbol in symbols:
                if _order_value(order.get("type")) == "trailing_stop":
                    continue
                # Calculate notional value of the open order
                qty = float(order.get("qty", 0) or 0)
                if order.get("side") == "sell":
                    qty = -qty  # Sell reduces position
                price = latest_prices.get(symbol, 0.0)
                notional = qty * price
                # Adjust weight
                if account.equity > 0:
                    weights_now[symbol] += notional / account.equity

    plans: list[OrderPlan] = []
    for symbol, target_weight in zip(symbols, target_weights, strict=True):
        current_weight = weights_now.get(symbol, 0.0)
        delta_weight = float(target_weight - current_weight)
        # Convert weight deltas into notional dollars so order sizing is tied
        # to portfolio equity instead of per-asset share math in this layer.
        notional_usd = abs(delta_weight) * account.equity
        # Ignore small drifts so the strategy does not churn on every run.
        if abs(delta_weight) < config.rebalance_threshold:
            continue
        if latest_prices.get(symbol, 0.0) <= 0.0:
            continue
        plans.append(
            OrderPlan(
                symbol=symbol,
                current_weight=round(current_weight, 6),
                target_weight=round(float(target_weight), 6),
                delta_weight=round(delta_weight, 6),
                side="buy" if delta_weight > 0 else "sell",
                notional_usd=round(notional_usd, 2),
            )
        )
    return plans


def build_trailing_stop_plan(
    *,
    symbols: list[str],
    target_weights: list[float],
    positions: list[Position],
    open_orders: list[dict] | None,
    trailing_stop: float,
    rebalance_threshold: float,
) -> TrailingStopPlanResult:
    target_by_symbol = {
        symbol: float(weight)
        for symbol, weight in zip(symbols, target_weights, strict=True)
    }
    protected_symbols = {
        str(order.get("symbol"))
        for order in open_orders or []
        if _order_value(order.get("type")) == "trailing_stop"
        and _order_value(order.get("side")) == "sell"
    }

    trail_percent = round(float(trailing_stop) * 100.0, 6)
    plans: list[TrailingStopPlan] = []
    unprotected_qty: list[UnprotectedTrailingStopQty] = []
    for position in positions:
        if position.symbol not in target_by_symbol:
            continue
        position_qty = float(position.qty)
        if position_qty <= 0.0:
            continue
        if target_by_symbol[position.symbol] < rebalance_threshold:
            continue
        if position.symbol in protected_symbols:
            continue
        whole_share_qty = float(floor(position_qty))
        remainder_qty = round(position_qty - whole_share_qty, 6)
        if remainder_qty > 0.0:
            unprotected_qty.append(
                UnprotectedTrailingStopQty(
                    symbol=position.symbol,
                    position_qty=round(position_qty, 6),
                    unprotected_qty=remainder_qty,
                )
            )
        if whole_share_qty >= 1.0:
            plans.append(
                TrailingStopPlan(
                    symbol=position.symbol,
                    qty=whole_share_qty,
                    side="sell",
                    trail_percent=trail_percent,
                    time_in_force="gtc",
                )
            )
    return TrailingStopPlanResult(orders=plans, unprotected_qty=unprotected_qty)


def _order_value(value: object) -> str:
    raw = getattr(value, "value", value)
    return str(raw)

File: rust-portfolio/src/portfolio_opt/rebalance.rs
use std::collections::HashMap;
use crate::portfolio_opt::types::{AccountSnapshot, OrderPlan, Position};
use crate::portfolio_opt::config::OptimizationConfig;

pub fn current_weights(
    symbols: &[String],
    account: &AccountSnapshot,
    positions: &[Position],
) -> HashMap<String, f64> {
    let by_symbol: HashMap<_, _> = positions.iter()
        .map(|p| (p.symbol.clone(), p))
        .collect();

    symbols.iter().map(|s| {
        let val = by_symbol.get(s)
            .map(|p| p.market_value)
            .unwrap_or(0.0);
        (s.clone(), if account.equity > 0.0 { val / account.equity } else { 0.0 })
    }).collect()
}

pub fn build_order_plan(
    symbols: &[String],
    target_weights: &[f64],
    account: &AccountSnapshot,
    positions: &[Position],
    latest_prices: &HashMap<String, f64>,
    config: &OptimizationConfig,
    open_orders: Option<&[serde_json::Value]>,
) -> Vec<OrderPlan> {
    let mut weights_now = current_weights(symbols, account, positions);

    // Adjust weights for pending orders
    if let Some(orders) = open_orders {
        for order in orders {
            if let (Some(symbol), Some(side)) = (order.get("symbol").and_then(|v| v.as_str()), order.get("side")) {
                if let Some(_pos) = symbols.iter().position(|s| s == symbol) {
                    let qty = order.get("qty").and_then(|v| v.as_f64()).unwrap_or(0.0);
                    let adjusted_qty = if side == "sell" { -qty } else { qty };
                    let price = latest_prices.get(symbol).copied().unwrap_or(0.0);
                    let notional = adjusted_qty * price;
                    if account.equity > 0.0 {
                        *weights_now.get_mut(symbol).unwrap() += notional / account.equity;
                    }
                }
            }
        }
    }

    let mut plans = Vec::new();
    for (symbol, &target_weight) in symbols.iter().zip(target_weights.iter()) {
        let current_weight = weights_now.get(symbol).copied().unwrap_or(0.0);
        let delta = target_weight - current_weight;
        let notional = delta.abs() * account.equity;

        if delta.abs() < config.rebalance_threshold {
            continue;
        }
        if latest_prices.get(symbol).copied().unwrap_or(0.0) <= 0.0 {
            continue;
        }

        plans.push(OrderPlan {
            symbol: symbol.clone(),
            current_weight,
            target_weight,
            delta_weight: delta,
            side: if delta > 0.0 { "buy".into() } else { "sell".into() },
            notional_usd: notional.round(),
        });
    }
    plans
}
